component/metrics/remotewrite/remote_write.go

```diff
@@ -0,0 +1,288 @@
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/storage/remote"
+)
+
+// Options.
+//
+// TODO(rfratto): These should be flags. How do we want to handle static
+// options for components?
+var (
+	walTruncateFrequency = 2 * time.Hour
+	minWALTime           = 5 * time.Minute
+	maxWALTime           = 8 * time.Hour
+	remoteFlushDeadline  = 1 * time.Minute
+	refCache             = wal.GlobalRefID
+)
+
+func init() {
+	remote.UserAgent = fmt.Sprintf("GrafanaAgent/%s", build.Version)
+	config.DefaultRemoteWriteConfig.SendExemplars = true
+
+	component.Register(component.Registration{
+		Name:    "metrics.remote_write",
+		Args:    RemoteConfig{},
+		Exports: Export{},
+		Build: func(o component.Options, c component.Arguments) (component.Component, error) {
+			return NewComponent(o, c.(RemoteConfig))
+		},
+	})
+
+}
+
+// RemoteConfig represents the input state of the metrics_forwarder component.
+type RemoteConfig struct {
+	ExternalLabels map[string]string    `hcl:"external_labels,optional"`
+	RemoteWrite    []*RemoteWriteConfig `hcl:"remote_write,block"`
+}
+
+// RemoteWriteConfig is the metrics_fowarder's configuration for where to send
+// metrics stored in the WAL.
+type RemoteWriteConfig struct {
+	Name      string           `hcl:"name,optional"`
+	URL       string           `hcl:"url"`
+	BasicAuth *BasicAuthConfig `hcl:"basic_auth,block"`
+}
+
+type Export struct {
+	Receiver *metrics.Receiver `hcl:"receiver"`
+}
+
+// BasicAuthConfig is the metrics_forwarder's configuration for authenticating
+// against the remote system when sending metrics.
+type BasicAuthConfig struct {
+	Username string          `hcl:"username"`
+	Password hcltypes.Secret `hcl:"password"`
+}
+
+// Component is the metrics_forwarder component.
+type Component struct {
+	log  log.Logger
+	opts component.Options
+	reg  *metrics.CollectorRegistry
+
+	walStore    *wal.Storage
+	remoteStore *remote.Storage
+	storage     storage.Storage
+
+	mut sync.RWMutex
+	cfg RemoteConfig
+
+	receiver *metrics.Receiver
+}
+
+// NewComponent creates a new metrics_forwarder component.
+func NewComponent(o component.Options, c RemoteConfig) (*Component, error) {
+	reg := metrics.NewCollectorRegistry()
+
+	// TODO(rfratto): don't hardcode base path
+	walLogger := log.With(o.Logger, "subcomponent", "wal")
+	dataPath := filepath.Join("data-agent", o.ID)
+	walStorage, err := wal.NewStorageWithRefIDSource(walLogger, reg, filepath.Join("data-agent", o.ID), refCache)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteLogger := log.With(o.Logger, "subcomponent", "rw")
+	remoteStore := remote.NewStorage(remoteLogger, reg, startTime, dataPath, remoteFlushDeadline, nil)
+
+	res := &Component{
+		log:  o.Logger,
+		opts: o,
+		reg:  reg,
+
+		walStore:    walStorage,
+		remoteStore: remoteStore,
+		storage:     storage.NewFanout(o.Logger, walStorage, remoteStore),
+	}
+	res.receiver = &metrics.Receiver{Receive: res.Receive}
```

Comment:
Set the receiver, I have to do it in a separate statement since it references itself.

Reply:
I suppose we can't have the function itself as the component's Export, right? It doesn't look so bad though.

Reply:
I actually didnt try that to see if cty can accept funcs, good call and I will check that.